lib/SILGen/SILGenStmt.cpp

```diff
@@ -932,10 +934,276 @@ void StmtEmitter::visitRepeatWhileStmt(RepeatWhileStmt *S) {
+                     std::move(result).ensurePlusOne(SGF, loc))
+          .forwardInto(SGF, initialization.get());
+    }
+  }
+
+  // If we ever reach an unreachable point, stop emitting statements.
+  // This will need revision if we ever add goto.
+  if (!SGF.B.hasValidInsertionPoint()) return;
+  
+  // If generator's optional result is address-only, create a stack allocation
+  // to hold the results.  This will be initialized on every entry into the loop
+  // header and consumed by the loop body. On loop exit, the terminating value
+  // will be in the buffer.
+  CanType optTy;
+  if (S->getConvertElementExpr()) {
+    optTy = S->getConvertElementExpr()->getType()->getCanonicalType();
+  } else {
+    optTy = OptionalType::get(S->getSequenceConformance().getTypeWitnessByName(
+                                  S->getSequence()->getType(),
+                                  SGF.getASTContext().Id_Element))
+                ->getCanonicalType();
+  }
+  auto &optTL = SGF.getTypeLowering(optTy);
+  SILValue addrOnlyBuf;
+  ManagedValue nextBufOrValue;
+
+  if (optTL.isAddressOnly() && SGF.silConv.useLoweredAddresses())
+    addrOnlyBuf = SGF.emitTemporaryAllocation(S, optTL.getLoweredType());
+
+  // Create a new basic block and jump into it.
+  JumpDest loopDest = createJumpDest(S->getBody());
+  SGF.B.emitBlock(loopDest.getBlock(), S);
+
+  // Compute the reference to the the generator's next() && cancel().
+  auto generatorProto =
+      SGF.getASTContext().getProtocol(KnownProtocolKind::AsyncIteratorProtocol);
+  ValueDecl *generatorNextReq = generatorProto->getSingleRequirement(
+      DeclName(SGF.getASTContext(), SGF.getASTContext().Id_next,
+               ArrayRef<Identifier>()));
+  ValueDecl *generatorCancelReq = generatorProto->getSingleRequirement(
+      DeclName(SGF.getASTContext(), SGF.getASTContext().Id_cancel,
+               ArrayRef<Identifier>()));
+  auto generatorAssocType =
+      asyncSequenceProto->getAssociatedType(SGF.getASTContext().Id_AsyncIterator);
+  auto generatorMemberRef = DependentMemberType::get(
+      asyncSequenceProto->getSelfInterfaceType(), generatorAssocType);
+  auto generatorType = sequenceConformance.getAssociatedType(
+      sequenceType, generatorMemberRef);
+  auto generatorConformance = sequenceConformance.getAssociatedConformance(
+      sequenceType, generatorMemberRef, generatorProto);
+  auto generatorSubs = SubstitutionMap::getProtocolSubstitutions(
+      generatorProto, generatorType, generatorConformance);
+  ConcreteDeclRef generatorNextRef(generatorNextReq, generatorSubs);
+  ConcreteDeclRef generatorCancelRef(generatorCancelReq, generatorSubs);
+
+  // Set the destinations for 'break' and 'continue'.
+  JumpDest endDest = createJumpDest(S->getBody());
+  SGF.BreakContinueDestStack.push_back({ S, endDest, loopDest });
+  
+
+  auto buildArgumentSource = [&]() {
+    if (cast<FuncDecl>(generatorNextRef.getDecl())->getSelfAccessKind() ==
```

Comment:
Is it not always mutating?

Reply:
it could perhaps not be, Iterators can be actors etc.